File: MB-Whatsapp/production-safety/env-validator.ts
import { EXPECTED_GLOBAL_AI_FLAGS, GLOBAL_AI_FLAGS_DEFAULTS, readFlag } from "./global-feature-flags";
import type { EnvValidationResult } from "./types";

export function validateProductionEnv(env: NodeJS.ProcessEnv = process.env): EnvValidationResult {
  const missing_flags = EXPECTED_GLOBAL_AI_FLAGS.filter((flag) => env[flag] === undefined);
  const dangerous_flags: string[] = [];
  const warnings: string[] = [];

  if (readFlag(env, "AI_RUNTIME_ENABLED", GLOBAL_AI_FLAGS_DEFAULTS.AI_RUNTIME_ENABLED) && !readFlag(env, "AI_RUNTIME_KILL_SWITCH", "true")) {
    dangerous_flags.push("ai_runtime_enabled_without_kill_switch");
  }
  if (readFlag(env, "MEDICAL_VISION_ENABLED", "false") && !readFlag(env, "HUMAN_REVIEW_ENABLED", "false")) {
    dangerous_flags.push("medical_vision_without_human_review");
  }
  if (readFlag(env, "CLINICAL_CONFIDENCE_BLOCKING_ENABLED", "false") && !readFlag(env, "HUMAN_REVIEW_ENABLED", "false")) {
    dangerous_flags.push("confidence_blocking_without_review");
  }
  if (readFlag(env, "OBSERVABILITY_EXTERNAL_EXPORT_ENABLED", "false") && readFlag(env, "OBSERVABILITY_PHI_ALLOWED", "false")) {
    dangerous_flags.push("observability_export_with_phi_allowed");
  }
  if (missing_flags.length > 0) warnings.push("expected_flags_missing_but_defaults_are_safe");

  return { ok: dangerous_flags.length === 0, missing_flags, dangerous_flags, warnings, checked_at: new Date().toISOString() };
}

File: api/app/runtime_support/production_safety/env_validator.py
from datetime import UTC, datetime
from collections.abc import Mapping

from .global_feature_flags import EXPECTED_GLOBAL_AI_FLAGS, GLOBAL_AI_FLAGS_DEFAULTS, read_flag
from .types import EnvValidationResult


def validate_production_env(env: Mapping[str, str]) -> EnvValidationResult:
    missing = [flag for flag in EXPECTED_GLOBAL_AI_FLAGS if flag not in env]
    dangerous: list[str] = []
    warnings: list[str] = []
    if read_flag(env, "AI_RUNTIME_ENABLED", GLOBAL_AI_FLAGS_DEFAULTS["AI_RUNTIME_ENABLED"]) and not read_flag(env, "AI_RUNTIME_KILL_SWITCH", "true"):
        dangerous.append("ai_runtime_enabled_without_kill_switch")
    if read_flag(env, "MEDICAL_VISION_ENABLED", "false") and not read_flag(env, "HUMAN_REVIEW_ENABLED", "false"):
        dangerous.append("medical_vision_without_human_review")
    if read_flag(env, "CLINICAL_CONFIDENCE_BLOCKING_ENABLED", "false") and not read_flag(env, "HUMAN_REVIEW_ENABLED", "false"):
        dangerous.append("confidence_blocking_without_review")
    if read_flag(env, "OBSERVABILITY_EXTERNAL_EXPORT_ENABLED", "false") and read_flag(env, "OBSERVABILITY_PHI_ALLOWED", "false"):
        dangerous.append("observability_export_with_phi_allowed")
    if missing:
        warnings.append("expected_flags_missing_but_defaults_are_safe")
    return EnvValidationResult(not dangerous, missing, dangerous, warnings, datetime.now(UTC).isoformat())
